src/libs/kata-sys-util/src/mount.rs

```diff
@@ -0,0 +1,1029 @@
+        Some("bind"),
+        MsFlags::MS_BIND,
+        Some(""),
+    )
+    .map_err(|e| Error::BindMount(abs_src, dst.to_path_buf(), e))?;
+
+    // Change into slave propagation mode.
+    mount(Some(""), dst, Some(""), MsFlags::MS_SLAVE, Some(""))
+        .map_err(|e| Error::Mount(PathBuf::new(), dst.to_path_buf(), e))?;
+
+    // Optionally rebind into readonly mode.
+    if read_only {
+        do_remount_read_only(dst, MsFlags::empty())?;
+    }
+
+    Ok(())
+}
+
+/// Trait to mount a `kata_types::mount::Mount`.
+pub trait Mounter {
+    /// Mount to the specified `target`.
+    ///
+    /// # Safety
+    /// Caller needs to ensure:
+    /// - `target` exists, and is suitable as destination for mount.
+    /// - `target` is free of file path based attacks.
+    fn mount<P: AsRef<Path>>(&self, target: P) -> Result<()>;
+}
+
+impl Mounter for kata_types::mount::Mount {
+    // This function is modelled after
+    // [Mount::Mount()](https://github.com/containerd/containerd/blob/main/mount/mount_linux.go)
+    // from [Containerd](https://github.com/containerd/containerd) project.
+    fn mount<P: AsRef<Path>>(&self, target: P) -> Result<()> {
+        fail::fail_point!("Mount::mount", |_| {
+            Err(Error::FailureInject(
+                "Mount::mount() fail point injection".to_string(),
+            ))
+        });
+
+        let target = target.as_ref().to_path_buf();
+        let (chdir, (flags, data)) =
+            // Follow the same algorithm as Containerd: reserve 512 bytes to avoid hitting one page
+            // limit of mounting argument buffer.
+            if self.fs_type == "overlay" && self.option_size() >= *MAX_MOUNT_PARAM_SIZE - 512 {
+                info!(
+                    sl!(),
+                    "overlay mount option too long, maybe failed to mount"
+                );
+                let (chdir, options) = compact_lowerdir_option(&self.options);
+                (chdir, parse_mount_options(&options)?)
+            } else {
+                (None, parse_mount_options(&self.options)?)
+            };
+
+        // Ensure propagation type change flags aren't included in other calls.
+        let propagation_types =
+            MsFlags::MS_SHARED | MsFlags::MS_PRIVATE | MsFlags::MS_SLAVE | MsFlags::MS_UNBINDABLE;
+        let o_flag = flags & (!propagation_types);
+
+        // - Normal mount without MS_REMOUNT flag
+        // - In the case of remounting with changed data (data != ""), need to call mount
+        if (flags & MsFlags::MS_REMOUNT) == MsFlags::empty() || !data.is_empty() {
+            mount_at(
+                chdir,
+                &self.source,
+                target.clone(),
+                &self.fs_type,
+                o_flag,
+                &data,
+            )?;
+        }
+
+        // Change mount propagation type.
+        if (flags & propagation_types) != MsFlags::empty() {
+            let propagation_flag = propagation_types | MsFlags::MS_REC | MsFlags::MS_SILENT;
```

Comment:
Worth a `debug!()` call here to show that this function changed the flags specified?

Reply:
Fixed